-1

Tested at: 5ebbc691492bc38eff41d049ab67eec9ea27db6e

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24365/29981/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTElectronDetaDphiProducer.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTCaloTowerProducer.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/ESListOfFEDSProducer.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc: In constructor 'EgammaHLTIslandClusterProducer::EgammaHLTIslandClusterProducer(const edm::ParameterSet&)':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc:71:67: error: no matching function for call to 'IslandClusterAlgo::IslandClusterAlgo(double, double, const PositionCalc&, IslandClusterAlgo::VerbosityLevel)'
                                        IslandClusterAlgo::pERROR)))
                                                                   ^
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/interface/EgammaHLTIslandClusterProducer.h:14:0,
                 from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc:34:
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-08-22-1100/src/RecoEcal/EgammaClusterAlgos/interface/IslandClusterAlgo.h:37:3: note: candidate: IslandClusterAlgo::IslandClusterAlgo(double, double, const PositionCalc&, const std::vector<int>&, const std::vector<int>&, const std::vector<int>&, const std::vector<int>&, IslandClusterAlgo::VerbosityLevel)
</pre>

--- a/Calibration/EcalCalibAlgos/interface/Pi0FixedMassWindowCalibration.h
+++ b/Calibration/EcalCalibAlgos/interface/Pi0FixedMassWindowCalibration.h
@@ -14,6 +14,7 @@
 #include "FWCore/Framework/interface/ESHandle.h"
 #include "FWCore/Framework/interface/Frameworkfwd.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/Framework/interface/EDProducer.h"
 #include "FWCore/Framework/interface/Event.h"
 #include "FWCore/Framework/interface/EventSetup.h"
@@ -65,6 +66,8 @@ class Pi0FixedMassWindowCalibration :  public edm::ESProducerLooper
   
   /// Destructor
   ~Pi0FixedMassWindowCalibration() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
 
   /// Dummy implementation (job done in duringLoop)
   virtual void produce(edm::Event&, const edm::EventSetup&) {};

--- a/Calibration/EcalCalibAlgos/src/Pi0FixedMassWindowCalibration.cc
+++ b/Calibration/EcalCalibAlgos/src/Pi0FixedMassWindowCalibration.cc
@@ -73,9 +73,6 @@ Pi0FixedMassWindowCalibration::Pi0FixedMassWindowCalibration(const edm::Paramete
   selePi0MinvMeanFixed_ = iConfig.getParameter<double>("selePi0MinvMeanFixed");
   selePi0MinvSigmaFixed_ = iConfig.getParameter<double>("selePi0MinvSigmaFixed");
 
-
-
-
   // Parameters for the position calculation:
   edm::ParameterSet posCalcParameters = 
     iConfig.getParameter<edm::ParameterSet>("posCalcParameters");
@@ -112,6 +109,58 @@ Pi0FixedMassWindowCalibration::Pi0FixedMassWindowCalibration(const edm::Paramete
 Pi0FixedMassWindowCalibration::~Pi0FixedMassWindowCalibration()
 {
 
+}
+
+void Pi0FixedMassWindowCalibration::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+
+  edm::ParameterSetDescription desc;
+
+  desc.add<unsigned int>("maxLoops", 0);
+  desc.add<std::string>("ecalRecHitsProducer", "");
+  desc.add<std::string>("barrelHitCollection", "");
+
+  desc.add<std::string>("VerbosityLevel", "");
+  desc.add<std::string>("barrelClusterCollection", "");
+
+  desc.add<double>("IslandBarrelSeedThr", 0);
+  desc.add<double>("IslandEndcapSeedThr", 0);
+
+  desc.add<double>("selePi0PtGammaOneMin", 0);
+  desc.add<double>("selePi0PtGammaTwoMin", 0);
+
+  desc.add<double>("selePi0DRBelt", 0);
+  desc.add<double>("selePi0DetaBelt", 0);
+
+  desc.add<double>("selePi0PtPi0Min", 0);
+
+  desc.add<double>("selePi0S4S9GammaOneMin", 0);
+  desc.add<double>("selePi0S4S9GammaTwoMin", 0);
+  desc.add<double>("selePi0S9S25GammaOneMin", 0);
+  desc.add<double>("selePi0S9S25GammaTwoMin", 0);
+
+  desc.add<double>("selePi0EtBeltIsoRatioMax", 0);
+
+  desc.add<double>("selePi0MinvMeanFixed", 0);
+  desc.add<double>("selePi0MinvSigmaFixed", 0);
+
+  edm::ParameterSetDescription posCalcParameters;
+  posCalcParameters.add<bool>("LogWeighted", true);
+  posCalcParameters.add<double>("T0_barl", 7.4);
+  posCalcParameters.add<double>("T0_endc", 3.1);
+  posCalcParameters.add<double>("T0_endcPresh", 1.2);
+  posCalcParameters.add<double>("W0", 4.2);
+  posCalcParameters.add<double>("X0", 0.89);
+  desc.add<edm::ParameterSetDescription>("posCalcParameters", posCalcParameters);
+
+  desc.add<std::string>("clustershapecollectionEB", "islandBarrelShape");
+  desc.add<std::string>("barrelShapeAssociation", "islandBarrelShapeAssoc");
+
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEE", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEE", {});
+
+  descriptions.add("Pi0FixedMassWindowCalibration", desc);
 }
 
 //_____________________________________________________________________________

--- a/Calibration/EcalCalibAlgos/src/Pi0FixedMassWindowCalibration.cc
+++ b/Calibration/EcalCalibAlgos/src/Pi0FixedMassWindowCalibration.cc
@@ -4,7 +4,6 @@
 
 // Framework
 
-
 // Conditions database
 
 #include "CondFormats/DataRecord/interface/EcalIntercalibConstantsRcd.h"
@@ -24,6 +23,8 @@
 // EgammaCoreTools
 #include "DataFormats/EgammaReco/interface/ClusterShape.h"
 #include "DataFormats/EgammaReco/interface/ClusterShapeFwd.h"
+
+#include "CommonTools/Utils/interface/StringToEnumValue.h"
 
 //const double Pi0Calibration::PDGPi0Mass =  0.1349766;
 
@@ -85,7 +86,19 @@ Pi0FixedMassWindowCalibration::Pi0FixedMassWindowCalibration(const edm::Paramete
   //AssociationMap
   barrelClusterShapeAssociation_ = iConfig.getParameter<std::string>("barrelShapeAssociation");
 
-  island_p = new IslandClusterAlgo(barrelSeedThreshold, endcapSeedThreshold, posCalculator_,verbosity);
+  const std::vector<std::string> seedflagnamesEB = iConfig.getParameter<std::vector<std::string> >("SeedRecHitFlagToBeExcludedEB");
+  const std::vector<int> seedflagsexclEB = StringToEnumValue<EcalRecHit::Flags>(seedflagnamesEB);
+
+  const std::vector<std::string> seedflagnamesEE = iConfig.getParameter<std::vector<std::string> >("SeedRecHitFlagToBeExcludedEE");
+  const std::vector<int> seedflagsexclEE = StringToEnumValue<EcalRecHit::Flags>(seedflagnamesEE);
+
+  const std::vector<std::string> flagnamesEB = iConfig.getParameter<std::vector<std::string> >("RecHitFlagToBeExcludedEB");
+  const std::vector<int> flagsexclEB = StringToEnumValue<EcalRecHit::Flags>(flagnamesEB);
+
+  const std::vector<std::string> flagnamesEE = iConfig.getParameter<std::vector<std::string> >("RecHitFlagToBeExcludedEE");
+  const std::vector<int> flagsexclEE = StringToEnumValue<EcalRecHit::Flags>(flagnamesEE);
+
+  island_p = new IslandClusterAlgo(barrelSeedThreshold, endcapSeedThreshold, posCalculator_, seedflagsexclEB, seedflagsexclEE, flagsexclEB, flagsexclEE, verbosity);
 
   theParameterSet=iConfig;
 

--- a/RecoEcal/EgammaClusterProducers/interface/IslandClusterProducer.h
+++ b/RecoEcal/EgammaClusterProducers/interface/IslandClusterProducer.h
@@ -31,7 +31,7 @@ class IslandClusterProducer : public edm::stream::EDProducer<>
       IslandClusterProducer(const edm::ParameterSet& ps);
 
       ~IslandClusterProducer() override;
-
+      static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
       void produce(edm::Event&, const edm::EventSetup&) override;
 
    private:

--- a/RecoEcal/EgammaClusterProducers/src/IslandClusterProducer.cc
+++ b/RecoEcal/EgammaClusterProducers/src/IslandClusterProducer.cc
@@ -9,6 +9,7 @@
 #include "DataFormats/Common/interface/Handle.h"
 #include "FWCore/Framework/interface/ESHandle.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/Utilities/interface/Exception.h"
 #include "CommonTools/Utils/interface/StringToEnumValue.h"
 
@@ -105,6 +106,36 @@ IslandClusterProducer::~IslandClusterProducer()
   delete island_p;
 }
 
+void IslandClusterProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+
+  edm::ParameterSetDescription desc;
+  desc.add<std::string>("VerbosityLevel", "ERROR");
+  desc.add<edm::InputTag>("barrelHits", edm::InputTag("ecalRecHit", "EcalRecHitsEB"));
+  desc.add<edm::InputTag>("endcapHits", edm::InputTag("ecalRecHit", "EcalRecHitsEE"));
+  desc.add<std::string>("barrelClusterCollection", "islandBarrelBasicClusters");
+  desc.add<std::string>("endcapClusterCollection", "islandEndcapBasicClusters");
+  desc.add<double>("IslandBarrelSeedThr", 0.5);
+  desc.add<double>("IslandEndcapSeedThr", 0.18);
+
+  edm::ParameterSetDescription posCalcParameters;
+  posCalcParameters.add<bool>("LogWeighted", true);
+  posCalcParameters.add<double>("T0_barl", 7.4);
+  posCalcParameters.add<double>("T0_endc", 3.1);
+  posCalcParameters.add<double>("T0_endcPresh", 1.2);
+  posCalcParameters.add<double>("W0", 4.2);
+  posCalcParameters.add<double>("X0", 0.89);
+  desc.add<edm::ParameterSetDescription>("posCalcParameters", posCalcParameters);
+
+  desc.add<std::string>("clustershapecollectionEE", "islandEndcapShape");
+  desc.add<std::string>("clustershapecollectionEB", "islandBarrelShape");
+  desc.add<std::string>("barrelShapeAssociation", "islandBarrelShapeAssoc");
+  desc.add<std::string>("endcapShapeAssociation", "islandEndcapShapeAssoc");
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEE", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEE", {});
+  descriptions.add("IslandClusterProducer", desc);
+}
 
 void IslandClusterProducer::produce(edm::Event& evt, const edm::EventSetup& es)
 {

--- a/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc
+++ b/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc
@@ -33,6 +33,8 @@
 // Class header file
 #include "RecoEgamma/EgammaHLTProducers/interface/EgammaHLTIslandClusterProducer.h"
 
+#include "CommonTools/Utils/interface/StringToEnumValue.h"
+
 EgammaHLTIslandClusterProducer::EgammaHLTIslandClusterProducer(const edm::ParameterSet& ps)
   : doBarrel_   (ps.getParameter<bool>("doBarrel"))
   , doEndcaps_  (ps.getParameter<bool>("doEndcaps"))
@@ -65,6 +67,10 @@ EgammaHLTIslandClusterProducer::EgammaHLTIslandClusterProducer(const edm::Parame
   , island_p (new IslandClusterAlgo(ps.getParameter<double>("IslandBarrelSeedThr"),
                                      ps.getParameter<double>("IslandEndcapSeedThr"),
                                      posCalculator_,
+                                     StringToEnumValue<EcalRecHit::Flags>(ps.getParameter<std::vector<std::string> >("SeedRecHitFlagToBeExcludedEB")),
+                                     StringToEnumValue<EcalRecHit::Flags>(ps.getParameter<std::vector<std::string> >("SeedRecHitFlagToBeExcludedEE")),
+                                     StringToEnumValue<EcalRecHit::Flags>(ps.getParameter<std::vector<std::string> >("RecHitFlagToBeExcludedEB")),
+                                     StringToEnumValue<EcalRecHit::Flags>(ps.getParameter<std::vector<std::string> >("RecHitFlagToBeExcludedEE")),
                                      verb_ == "DEBUG" ? IslandClusterAlgo::pDEBUG :
                                      (verb_ == "WARNING" ? IslandClusterAlgo::pWARNING :
                                       (verb_ == "INFO" ? IslandClusterAlgo::pINFO :

--- a/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc
+++ b/RecoEgamma/EgammaHLTProducers/src/EgammaHLTIslandClusterProducer.cc
@@ -107,6 +107,10 @@ void EgammaHLTIslandClusterProducer::fillDescriptions(edm::ConfigurationDescript
   desc.add<double>("regionEtaMargin", 0.3);
   desc.add<double>("regionPhiMargin", 0.4);
   //desc.add<edm::ParameterSet>("posCalcParameters"), edm::ParameterSet());
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("SeedRecHitFlagToBeExcludedEE", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEB", {});
+  desc.add<std::vector<std::string>>("RecHitFlagToBeExcludedEE", {});
   descriptions.add("hltEgammaHLTIslandClusterProducer", desc);  
 }
 

